components/style/gecko_selector_impl.rs

```diff
@@ -12,104 +12,85 @@ use stylesheets::Stylesheet;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct GeckoSelectorImpl;
 
+/// NOTE: The boolean field represents whether this element is an anonymous box.
+///
+/// This is just for convenience, instead of recomputing it. Also, note that
+/// Atom is always a static atom, so if space is a concern, we can use the
+/// raw pointer and use the lower bit to represent it without space overhead.
+///
+/// FIXME(emilio): we know all these atoms are static. Patches are starting to
+/// pile up, but a further potential optimisation is generating bindings without
+/// `-no-gen-bitfield-methods` (that was removed to compile on stable, but it no
+/// longer depends on it), and using the raw *mut nsIAtom (properly asserting
+/// we're a static atom).
+///
+/// This should allow us to avoid random FFI overhead when cloning/dropping
+/// pseudos.
+///
+/// Also, we can further optimize PartialEq and hash comparing/hashing only the
+/// atoms.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum PseudoElement {
-    Before,
-    After,
-
-    Backdrop,
-    FirstLetter,
-    FirstLine,
-    MozSelection,
-    MozFocusInner,
-    MozFocusOuter,
-    MozListBullet,
-    MozListNumber,
-    MozMathAnonymous,
-    MozNumberWrapper,
-    MozNumberText,
-    MozNumberSpinBox,
-    MozNumberSpinUp,
-    MozNumberSpinDown,
-    MozProgressBar,
-    MozRangeTrack,
-    MozRangeProgress,
-    MozRangeThumb,
-    MozMeterBar,
-    MozPlaceholder,
-    MozColorSwatch,
-
-    AnonBox(AnonBoxPseudoElement),
-}
+pub struct PseudoElement(Atom, bool);
```

Comment:
It's documented above (though the diff makes it difficult to find), but
yeah, it's fair, I'll do it.

On 08/13/2016 08:04 AM, Simon Sapin wrote:
> In components/style/gecko_selector_impl.rs
> <https://github.com/servo/servo/pull/12815#discussion_r74687932>:
> 
>> -    MozNumberWrapper,
>> -    MozNumberText,
>> -    MozNumberSpinBox,
>> -    MozNumberSpinUp,
>> -    MozNumberSpinDown,
>> -    MozProgressBar,
>> -    MozRangeTrack,
>> -    MozRangeProgress,
>> -    MozRangeThumb,
>> -    MozMeterBar,
>> -    MozPlaceholder,
>> -    MozColorSwatch,
>> -
>> -    AnonBox(AnonBoxPseudoElement),
>> -}
>> +pub struct PseudoElement(Atom, bool);
> 
> Maybe use struct syntax with named fields? I particular to say what this
> |bool| is.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/12815/files/8a92235c7c372b7d884a6928608a7477e2af7c70#r74687932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQwull1-KtkBA4749IbfIn2nU6g_1AUks5qfd0PgaJpZM4Jh0-a>.
>